mailpoet/lib/Automation/Integrations/Core/Actions/Wait/Action.php

```diff
@@ -0,0 +1,60 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Integrations\Core\Actions\Wait;
+
+use MailPoet\Automation\Engine\Control\ActionScheduler;
+use MailPoet\Automation\Engine\Hooks;
+use MailPoet\Automation\Engine\Workflows\ActionInterface;
+use MailPoet\Automation\Engine\Workflows\Step;
+use MailPoet\Automation\Engine\Workflows\Workflow;
+use MailPoet\Automation\Engine\Workflows\WorkflowRun;
+use MailPoet\InvalidStateException;
+
+class Action implements ActionInterface {
+  /** @var ActionScheduler */
+  private $actionScheduler;
+
+  public function __construct(
+    ActionScheduler $actionScheduler
+  ) {
+    $this->actionScheduler = $actionScheduler;
+  }
+
+  public function getKey(): string {
+    return 'core:wait';
+  }
+
+  public function getName(): string {
+    return __('Wait', 'mailpoet');
+  }
+
+  public function run(Workflow $workflow, WorkflowRun $workflowRun, Step $step): void {
+    $result = $this->validate($workflow, $step);
+    if (!$result->isValid()) {
+      throw InvalidStateException::create()->withErrors($result->getErrors());
+    }
+    $this->actionScheduler->schedule(time() + $result->getWaitTime(), Hooks::WORKFLOW_STEP, [
+      [
+        'workflow_run_id' => $workflowRun->getId(),
+        'step_id' => $step->getNextStepId(),
+      ],
+    ]);
+
+    // TODO: call a step complete ($id) hook instead?
+  }
+
+  public function validate(Workflow $workflow, Step $step, array $subjects = []): ValidationResult {
```

Comment:
It looks like [covariance](https://www.php.net/manual/en/language.oop5.variance.php) (return types more specific than a parent's) isn't fully supported until 7.4, so this is causing static analysis to [fail](https://app.circleci.com/pipelines/github/mailpoet/mailpoet/9460/workflows/51fbb356-cce7-4f27-88e0-b5956fa8960b/jobs/173900). I would really like to be able to do this as written rather than defining the return type as `AbstractValidationResult` and then having to check if it's an instance of the more specific class that I know that it is. Is there a better way to handle this in a PHP 7.2-friendly way?

Reply:
Maybe PHPStan and IDEs would understand if you add `/** @return MyMoreSpecificType */` (keeping the PHP-native return type more general).